README.rst
.. _Recalbox: http://recalbox.com
.. _recalbox-webconfig: https://github.com/MikaXII/recalbox-webconfig
.. _Django: https://www.djangoproject.com
.. _Foundation: http://foundation.zurb.com
.. _autobreadcrumbs: https://github.com/sveetch/autobreadcrumbs
.. _virtualenv: http://www.virtualenv.org/
.. _psutil: https://pypi.python.org/pypi/psutil
.. _Dropzone.js: http://www.dropzonejs.com/

Recalbox manager Web interface
==============================

Like `recalbox-webconfig`_ this project aims to serve a web interface to manage some common `Recalbox`_ configurations but with `Django`_ instead of *Node.js*.

This is a full Django webapp project, meaning it's ready to launch when correctly installed.

Features
********

* Try to be the lightweight as possible (even if using Django):
    
  * Actually don't have any app models, so we never perform database request for anything;
  * Don't use Django site framework (to avoid database request);
  * Don't use compressor system (like django-assets or django-pipeline) to avoid processing many files just to display static files tags. Instead ships allready compressed assets and switch to them in production environment;

* Hardly repose on Recalbox Manifest file to validate uploads;
* Web integration on top of `Foundation`_;
* Display system informations like CPU, Memory and disks usage;
* Read the Recalbox logs;
* Edit the Recalbox configuration file;
    
  * Option to backup the file before updating it;

* Manage (upload, delete) your roms by systems;
  
  * Only accept supported extensions for systems (from manifest);
  
* Manage (upload, delete) your bios files;

  * Only accept supported Bios file (from manifest);
  * MD5 checksum validation;

* Usage of `Dropzone.js`_ plugin for upload forms;
  

Install
*******

Common Linux system (development)
---------------------------------

Nothing special, it's just about to have PIP and `virtualenv`_ installed on your system, enter into your recalbox-manager directory, then use the Makefile action: ::

    make install

And voila, it's done.

But note this procedure is mostly for development purpose. See next section.

Recalbox system (production)
----------------------------

Recalbox system is assumed to be the production environment.

This is different because Recalbox don't have all libraries and tools common to Linux systems.

Before doing anything, ensure the Raspberry can access to the internet else configure your network interface and if needed dns resolving.

Go where you want to install the manager directory then type the following commands: ::

    wget -q -O - https://raw.githubusercontent.com/sveetch/recalbox-manager/master/deployment/install.sh | bash /dev/stdin --release=1.1.4.1

This will download an install script and automatically execute it to proceed to install.

Actually the install script accept an additional argument to enable compatibility with Recalbox 3.2.11 (because it's actually the real last stable version). So to install recalbox-manager on Recalbox 3.2.11, add `` --compatible`` to the previous command.

Usage
*****

Basically, go into the recalbox-manager directory then use the following commands: ::

    bin/python manage.py runserver 0.0.0.0:8001

The runserver is not launched as a daemon or a background process, as soon as you stop the instance thread (using CTRL-C) the webserver is stopped.

**For production** you must use the right settings: ::

    bin/python manage.py runserver 0.0.0.0:80 --settings=project.settings_production --noreload

**For development**, you should active the environment and you may want to use the right settings with additional stuff: ::

    . bin/activate
    python manage.py runserver 0.0.0.0:8001 --settings=project.settings_development
    
Notes for development
*********************

#. Before doing anything (install or whatever), development requires additional tools : *Ruby*, *Node.js* and *npm*. Install them on your system;

#. To directly install the full development environment, just use ``make install-dev`` from the project root, it will install everything (use ``make clean`` before if you previously used the ``make install`` command) but the Compass stuff.

#. The ``pip-requirements/development.txt`` contains some additional packages;

#. You can install the project on common Linux system for development but you will need to reproduce the Recalbox file structure for Roms, Bios, Configuration file, log file, etc.. Or you can edit needed paths in project settings;

#. CSS are compiled from Compass sources, you will need to install the right Compass (use the shipped ``Gemfile`` file) and Foundation 5 (use the dedicated Makefile action) versions;

#. Python 2.7.9 is installed on Recalbox 3.2.11, so *pip* is near to be ready to use, just have to install it the first time. This will results to install ``pip==1.5.6``.

Assets
------

You need to install the required Grunt stuff to develop on assets, it should have been done with ``make install-dev``

Assets are managed in a JSON manifest ``project/assets.json`` that are used by Django template tags to know what asset to load in the pages. And the manifest is used also by Grunt tasks to optimize and build the asset files for production environment. 

In default and development environment loaded assets are not uglified or compressed to ease asset debugging.

When you did some changes (add, delete, change) on Javascript files, you will need to execute the following Grunt task: ::

    grunt uglify

And when you did some changes on CSS files (or when Compass rebuild CSS from your SCSS changes), you will need to execute the following Grunt task: ::

    grunt cssmin

Also to make continue development, you can use the watch task so every time Compass is making a recompile, cssmin will compress CSS: ::

    grunt watch

**Remember** to execute theses tasks before commiting updates on assets.

Notes for production
********************

Last tests on Recalbox 3.3.0 beta 6 and recalbox-manager==0.8.2 was giving 2% CPU charge when Django instance is idle and can go to 17% when furiously reloading a page during 30seconds. Memory is allways stable around 80Mo and should probably don't go further. This was a naive benchmark just using ``top``.

Caveats
*******

* Python devel lib is not installed on Recalbox, this would prevent you to be able to install somes additional Python packages that require to compile some C code;

* Currently, webapp is served using the development server from Django. It's awful but at least the webapp should not have to response to many connections. This choice has been done to avoid loading a real web server on the Raspberry additionaly to the Django instance;

* UTC Timezone does not seems available for now, so have to set ``settings.TIME_ZONE`` to ``None`` and set ``settings.USE_TZ`` to ``False``;


HISTORY.rst
History
=======

Version 1.1.4.1 - 2015/12/22
----------------------------

* Updated PO and french translation;
* Minor improvement on some CSS stuff;

Version 1.1.4 - 2015/12/21
--------------------------

* Added CPU temperature in monitoring using ACPI thermal API, close #38;
* Notify about missing system bios in rom list view, close #50;

Version 1.1.3 - 2015/12/19
--------------------------

* Updated Manifest to add zip extensions on some systems, related to #44 and #46;
* Granted ``bin`` extension for roms on Atari2600 system;
* Updated basic pip requirement file to fix Django version on 1.8.x for now;
* Fixed Makefile action ``install`` to use the right pip requirement file path;
* Added a global page footer containing the manager version, close #45;
* Only display supported extensions for knowed rom systems from manifest, for unknow system ddirs don't filter on file extension, close #40;
* Unmount ``/admin/`` because it's useless and may give back some ressources;
* Naive fix to avoid throwing exception on bad encoded filename in rom list view, just ignore it and continue. Related to #39;

Version 1.1.1 - 2015/11/01
--------------------------

* Cleaning deprecated stuff in assets_cartographer;
* Minor changes on url/texts in homepage;
* Added missing translation blocks;
* Updated PO and french translation;

Version 1.1.0 - 2015/10/17
--------------------------

* Better 'manager_frontend' structure, cleaning some minor things, close #34;
* Added ``asset_tag``;
* Dont show 'available systems' form if all systems allready exist, close #35;
* Moved 'Recalbox manifest' stuff into ``project.recalbox_manifest``, removed manifest loading from settings in profit of loading from ``urls.py``;

Version 1.0.2 - 2015/10/16
--------------------------

* Fixed CRLF caracters when editing recalbox config, close #31;
* Fixed not packaged assets with production settings;
* Refactored asset management to use persistent manifest registry (avoid to load JSON manifest file on every request) and move it to its own embedded app to ``project.assets_cartographer``, close #33;
* Open Virtual gamepad link into a new window/tab;
* Fixed Dropzone assets, close #32;
* Moved Python requirements to ``pip-requirements`` directory and refactored them for backward compatible with Recalbox versions without ``psutil`` yet;
* Added automatic deployment bash script in ``deployment/install.sh``, update Readme install procedure;

Version 1.0.1 - 2015/10/14
--------------------------

* Fixed wrong port for Virtual gamepad to 8080 (instead of 8081);

Version 1.0.0 - 2015/10/11
--------------------------

* Added Dropzone.js asset;
* Added all machinary to enabled Dropzone on rom upload;

Version 0.9.0 - 2015/10/03
--------------------------

* Added django migration command to makefile action ``install``;
* Added makefile action ``install-dev`` to install additional stuff for development;
* Added makefile action ``assets`` to build optimized assets (like for before committing asset changes);
* Moved ``psutil`` dependancies in basic ``requirements.txt``;
* Don't load anymore the whole Foundation components, only the used ones (win ~80ko on assets);
* Added homemade Asset management:
    
  * Centralized asset definition in a manifest file in ``project/assets.json``;
  * Use a template tag to read asset manifest to know what files to load for stylesheets and Javascripts assets;
  * Added Grunt stuff to use the asset manifest to build their optimized version for production;

Version 0.8.4 - 2015/09/26
--------------------------

* Updated dev requirements to ``django-icomoon==0.2``;
* Added ``hostname`` var to ``SITE`` in template context;
* Updated webfont;
* Added link to Recalbox 'virtual gamepad' app on homepage, close #24;
* Update README.rst;
* Minor fix in readme for runserver command for production;


Version 0.8.3 - 2015/09/21
--------------------------

* Updated package url in README
* Fixed Config edition error, close #18;
* Removed deprecated setting ``TEMPLATE_DEBUG``, close #20;
* Moved system information to its own view to avoid loading time on homepage, close #19;
* Fixed missing Foundation Javascripts assets, close #21;


Version 0.8.2 - 2015/09/20
--------------------------

* Fixed Rom systems grid using flexbox;
* Dont display System infos part on homepage if ``psutil`` is not installed;
* Added ``settings.RECALBOX_PSUTIL_CPU_INTERVAL`` to define time blocking interval to watch for cpu usage;

Version 0.8.1 - 2015/09/20
--------------------------

* Updated PO for last translatable string;
* Updated french PO;

Version 0.8.0 - 2015/09/20
--------------------------

* Added requirement file for development and Recalbox 3.3.0 beta5;
* Added ``django-icomoon usage`` for development;
* Added settings_production.py to use port 80 in production environment, close #13;
* Added safe usage of ``psutil`` library in homepage to display system infos, close #16;
* Removed Django contrib ``Site framework`` usage because we don``t use it anymore, related to #6;

Version 0.7.1 - 2015/09/06
--------------------------

* Finally finded a proper way to find host ip
* Added some settings so we can fix some host infos, related to #13
* Continued on system infos mining but disabled it (seems a fail)
* Added flexbox stuff in scss


Version 0.7.0 - 2015/09/06
--------------------------

* Added translated link to Recalbox wiki, #14;
* Removed template context variable ``SITE.domain`` usage in skeleton, #13;
* Updated Readme to include a line about migration on install, #12;
* Added settings_development and moved debug_toolbar instructions into this new settings env, #11;
* Updated PO files;
* Started to get system infos to display on homepage;
* Fix syntax error in German PO file, compile PO;
* Added German PO file;

Version 0.6.0 - 2015/08/31
--------------------------

* Parse the XML manifest within settings, close #3;
* Added XML manifest parser and (temporary?) ship the XML file for issue #3

Version 0.5.0 - 2015/08/31

project/settings.py
# -*- coding: utf-8 -*-
"""
Django settings for recalbox-manager project.

Generated by 'django-admin startproject' using Django 1.8.3.

For more information on this file, see
https://docs.djangoproject.com/en/1.8/topics/settings/

For the full list of settings and their values, see
https://docs.djangoproject.com/en/1.8/ref/settings/
"""

# Build paths inside the project like this: os.path.join(BASE_DIR, ...)
import os
#from manager_frontend.utils.manifest import RecalboxManifestParser

gettext = lambda s: s

BASE_DIR = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
PROJECT_DIR = os.path.join(BASE_DIR, 'project')

SITE_ID = 1
# If not empty, use this instead of Site framework to know the name and use 
# a trick to find the host ip
SITE_FIXED = {
    'name': "Recalbox Manager",
    'ip': None, # If None find the ip automatically, else use a string to define another hostname
    'port': '8001', # If None no port is appended to hostname, so the server have to be reachable from port 80
}

# Quick-start development settings - unsuitable for production
# See https://docs.djangoproject.com/en/1.8/howto/deployment/checklist/

# SECURITY WARNING: keep the secret key used in production secret!
SECRET_KEY = '(_0!&^^xekahfp=s5(9+^wlq6gvn6z90%i*p+wn^4ir+mvl4lx'

# SECURITY WARNING: don't run with debug turned on in production!
DEBUG = True

ALLOWED_HOSTS = []


# Application definition

INSTALLED_APPS = (
    'django.contrib.admin',
    'django.contrib.auth',
    'django.contrib.contenttypes',
    'django.contrib.sessions',
    'django.contrib.messages',
    #'django.contrib.sites',
    'django.contrib.staticfiles',
    
    'autobreadcrumbs',
    
    'project.assets_cartographer',
    'project.recalbox_manifest',
    'project.manager_frontend',
    
    #'ajaxuploader',
)

MIDDLEWARE_CLASSES = (
    'django.contrib.sessions.middleware.SessionMiddleware',
    'django.middleware.locale.LocaleMiddleware',
    'django.middleware.common.CommonMiddleware',
    'django.middleware.csrf.CsrfViewMiddleware',
    'django.contrib.auth.middleware.AuthenticationMiddleware',
    'django.contrib.auth.middleware.SessionAuthenticationMiddleware',
    'django.contrib.messages.middleware.MessageMiddleware',
    #'django.contrib.sites.middleware.CurrentSiteMiddleware',
    'django.middleware.clickjacking.XFrameOptionsMiddleware',
    'django.middleware.security.SecurityMiddleware',
)

ROOT_URLCONF = 'project.urls'

TEMPLATES = [
    {
        'BACKEND': 'django.template.backends.django.DjangoTemplates',
        'DIRS': [
            os.path.join(PROJECT_DIR, "templates"),
        ],
        'APP_DIRS': True,
        'OPTIONS': {
            'context_processors': [
                'django.template.context_processors.debug',
                'django.template.context_processors.i18n',
                'django.template.context_processors.request',
                'django.template.context_processors.static',
                'django.template.context_processors.media',
                'django.contrib.auth.context_processors.auth',
                'django.contrib.messages.context_processors.messages',
                'project.utils.context_processors.site_metas',
                'project.utils.context_processors.manager_version',
                'autobreadcrumbs.context_processors.AutoBreadcrumbsContext',
            ],
        },
    },
]

WSGI_APPLICATION = 'project.wsgi.application'


# Database
# https://docs.djangoproject.com/en/1.8/ref/settings/#databases
DATABASES = {
    'default': {
        'ENGINE': 'django.db.backends.sqlite3',
        'NAME': os.path.join(BASE_DIR, 'db.sqlite3'),
    }
}


# Internationalization
# https://docs.djangoproject.com/en/1.8/topics/i18n/

LANGUAGE_CODE = 'en-us'

LANGUAGES = (
    ('en', gettext('English')),
    ('fr', gettext('French')),
    ('de', gettext('German')),
)

TIME_ZONE = None

USE_I18N = True

USE_L10N = True

USE_TZ = False

LOCALE_PATHS = (
    os.path.join(PROJECT_DIR, 'locale'),
)

# Absolute filesystem path to the directory that will hold user-uploaded files.
# Example: "/var/www/example.com/media/"
MEDIA_ROOT = os.path.join(PROJECT_DIR, 'media')

# URL that handles the media served from MEDIA_ROOT. Make sure to use a
# trailing slash.
# Examples: "http://example.com/media/", "http://media.example.com/"
MEDIA_URL = '/media/'

# Static files (CSS, JavaScript, Images)
# https://docs.djangoproject.com/en/1.8/howto/static-files/
STATIC_ROOT = os.path.join(PROJECT_DIR, 'static')
STATIC_URL = '/static/'
STATICFILES_DIRS = (
    os.path.join(PROJECT_DIR, "webapp_statics"),
)

# For Django messages framework to avoid database requests
MESSAGE_STORAGE = 'django.contrib.messages.storage.cookie.CookieStorage'

# For assets management
ASSETS_PACKAGED = not(DEBUG)
ASSETS_STRICT = False
ASSETS_MAP_FILEPATH = os.path.join(PROJECT_DIR, "assets.json")
ASSETS_TAG_TEMPLATES = {
    "javascripts": "assets/javascript_tag.html",
    "stylesheets": "assets/stylesheet_tag.html",
}

#
# Recalbox needed paths
#

# Path to directory that contains bios file
RECALBOX_BIOS_PATH = "/recalbox/share/bios"
# Path to directory that contains system roms directories
RECALBOX_ROMS_PATH = '/recalbox/share/roms'
# Path to the Recalbox configuration file
RECALBOX_CONF_PATH = '/recalbox/share/system/recalbox.conf'
# Path to the Recalbox configuration backup file
RECALBOX_CONF_BACKUP_PATH = '/recalbox/share/system/recalbox.old.conf'
# Path to the Recalbox logs file
RECALBOX_LOGFILE_PATH = "/recalbox.log"
# Path to the Recalbox manifest file (actually shipped into manager project)
RECALBOX_MANIFEST_FILEPATH = os.path.join(PROJECT_DIR, 'MANIFEST.xml')

# Default empty entry for unknowed system from Manifest
RECALBOX_SYSTEM_DEFAULT = {
    'name': 'Unknow',
    'extensions': [],
    'download_links': [],
    'bios': [],
}

# Blocking time during psutil watch for cpu charge in second (float number)
RECALBOX_PSUTIL_CPU_INTERVAL = 0.5 # 0.1 seems a little too low but 1.0 add 1s on page loading time
